server/src/main/java/org/apache/iotdb/db/qp/strategy/optimizer/ConcatPathOptimizer.java

```diff
@@ -238,15 +239,101 @@ private FilterOperator concatFilter(List<Path> fromPaths, FilterOperator operato
+    // brings two 'time < 10' filter operator
+    Set<FilterOperator> operatorSet = new HashSet<>();
+    for (int i = 0; i < noStarDevices.size(); i++) {
+      FilterOperator newOperator = operator.clone();
+      newOperator = concatFilterPath(noStarDevices.get(i), newOperator);
+
+      if (!operatorSet.contains(newOperator)) {
+        if(currentNode.getChildren().size() > 0){
+          FilterOperator newInnerNode = new FilterOperator(SQLConstant.KW_OR);
+          currentNode.addChildOperator(newInnerNode);
+          parentNode = currentNode;
+          currentNode = newInnerNode;
+        }
+        currentNode.addChildOperator(newOperator);
+        operatorSet.add(newOperator);
+      }
+    }
+
+    // if 'OR' has no enough operands due to duplication
+    if(currentNode.getChildren().size() == 1){
+      if(parentNode == currentNode){
+        filterBinaryTree = currentNode.getChildren().get(0);
+      } else {
+        parentNode.getChildren().set(1, currentNode.getChildren().get(0));
+      }
+    }
+
+    return filterBinaryTree;
+  }
+
+  private List<String> removeStarsInDeviceWithUnique(List<Path> paths)
+          throws LogicalOptimizeException {
+    List<String> retDevices;
+    Set<String> deviceSet = new LinkedHashSet<>();
+    try {
+      for (Path path : paths) {
+        List<String> tempDS;
+        tempDS = MManager.getInstance().getDevices(path.getFullPath());
+
+        for (String subDevice : tempDS) {
+          if (!deviceSet.contains(subDevice)) {
+            deviceSet.add(subDevice);
+          }
+        }
+      }
+      retDevices = new ArrayList<>(deviceSet);
+    } catch (PathException e) {
+      throw new LogicalOptimizeException("error when remove star: " + e.getMessage());
     }
+    return retDevices;
+  }
+
+  private FilterOperator concatFilterPath(String prefix, FilterOperator operator) {
+    if(!operator.isLeaf()){
+      for (FilterOperator child : operator.getChildren()) {
+        concatFilterPath(prefix, child);
+      }
+      return operator;
+    }
+    BasicFunctionOperator basicOperator = (BasicFunctionOperator) operator;
+    Path filterPath = basicOperator.getSinglePath();
+
+    // do nothing in the cases of "where time > 5" or "where root.d1.s1 > 5"
+    if (SQLConstant.isReservedPath(filterPath) || filterPath.startWith(SQLConstant.ROOT)) {
+      return operator;
+    }
+
+    Path concatPath = filterPath.addPrefixPath(filterPath, prefix);
+    basicOperator.setSinglePath(concatPath);
+
+    return basicOperator;
   }
 
-  private FilterOperator constructTwoForkFilterTreeWithAnd(List<Path> noStarPaths,
+  private FilterOperator constructBinaryFilterTreeWithAnd(List<Path> noStarPaths,
       FilterOperator operator)
       throws LogicalOptimizeException {
-    FilterOperator filterTwoFolkTree = new FilterOperator(SQLConstant.KW_AND);
-    FilterOperator currentNode = filterTwoFolkTree;
+    FilterOperator filterBinaryTree = new FilterOperator(SQLConstant.KW_AND);
```

Comment:
```suggestion
    FilterOperator binaryFilter = new FilterOperator(SQLConstant.KW_AND);
```

Reply:
Keep it. It's better to indicate what we are doing.